Split release checklist into a checklist and steps (#3987)
unicode-org/icu4x: docs/process/release.md

Over the last month before the release target date, we'll discuss the checklist and milestone progress in our weekly calls, and fine tune timing and scope.

## Pre-release checklist

This is a checklist of things that should be done in the weeks leading to the release.

* [ ] Verify that the milestone and checklist are complete
* [ ] Verify with component owners that they're ready for release
* [ ] Take a bird-eye view at:
  * [ ] READMEs
  * [ ] Documentation
  * [ ] Coverage
  * [ ] Performance / Memory / Size benchmarks
  * [ ] Cargo.toml files
    * [ ] All dependencies from the workspace should use `workspace = true` rather than their version number or path 

>     * [ ] Cargo.toml files need to specify versions for each non-workspace entry in `dependencies`

Reviewer: do we want to use workspace dependencies for non-workspace dependencies as well? I've only done workspace dependencies so far because that's a low-hanging fruit

Author: We can give people the option

Reviewer: I think we should do either all or nothing

Author: I think it's fine for super common deps but overkill for everything. Would support a gradual move towards this for most deps.

>     * [ ] Cargo.toml should not specify versions for each workspace entry in `dev-dependencies`.

Reviewer: not relevant with `workspace = true`
```suggestion
```

>     * [ ] Ensure that any new packages have suitable version numbers.

Reviewer: ```suggestion
    * [ ] Ensure that any new packages have version number `0.0.0`, this will making bumping during the release easier.
```

Author: Added, but I'm not a fan of having the process be overly dependent on the tool.

Reviewer: Nothing to do with the tool, even when bumping version numbers manually, it's clear that `0.0.0` has never been released but `0.1.0` could have been.

Author: Fair.

>     * [ ] Ensure that the Cargo.toml file includes a set of fields consistent with other ICU4X packages.

Reviewer: ```suggestion
    * [ ] Ensure that the Cargo.toml file includes a set of fields consistent with other ICU4X packages.
        * These should mostly use workspace inheritance
```

* [ ] Run `cargo +nightly fmt -- --config=format_code_in_doc_comments=true --config=doc_comment_code_block_width=80` to prettify our docs

> * [ ] Run `cargo update` for each `Cargo.lock` file to update our CI to freshest dependencies

Reviewer: These should happen during the release, not weeks before.

Author: I think it's fine to preemptively do a run. The release checklist involves rerunning these.

* [ ] Go through `ffi/diplomat/tests/missing_apis.txt` and verify that it is empty. If it is not, component owners should either add FFI APIs, add `rust_link` annotations, or allowlist the relevant APIs as having been punted to the future
* [ ] Verify that `ffi/diplomat` depends on a released (not Git) version of Diplomat. Get it published (ask manishearth or sffc) otherwise.

> * [ ] Start work on a changelog (see below)

Reviewer: ```suggestion
* [ ] Start work on a changelog (see below)
* [ ] Prepare a PR to update tutorials using the upcoming release. The PR should pass `cargo make test-tutorials-local`, but can fail `cargo make test-tutorials-cratesio` prior to release
```

## Release steps

Once the release checklist is complete, the assigned release driver will perform the following steps, in order:

> * Go through the prerelease checklist again, ensuring that no problems were reintroduced in the PRs that landed since the opening of the checklist. (Things like doc prettification will likely need to be rerun!)

Reviewer: why rerun if we can just run it once? it's a lot of churn

Author: Ideally it will be a small amount of churn the second time. I think it's good to preflight as much as possible.

Author: Basically most of that checklist is going to need to be rerun anyway; because the whole point of this PR is to give people a bunch of parallelizable work that makes the final release smoother and less contentious, with the understanding that the parallelizeable work will happen over the course of a couple weeks and may miss changes that happen later.

* Release utils (see section below). Get the version bumps reviewed and checked in before releasing.
* Update ICU4X package versions as needed. Most of this can be done by updating `workspace.package` in the root `Cargo.toml` and the `workspace.dependencies` entries there. Some `icu_*` crates do not follow the ICU4X versioning scheme like `icu_codepointtrie_builder` or experimental crates.
* Temporarily add `icu_testdata` back to the workspace
  * To build, run `cargo make testdata-legacy-gen` to generate the gitignored data
* Get this reviewed and checked in before continuing
* Publish each `icu_*` crate
  * Use `cargo workspaces publish --from-git` to automatically publish the crates in the correct order if you would like
  * Add `icu4x-release` group as owners to each new component you're publishing
    * `cargo owner -a github:unicode-org:icu4x-release`

> * Verify that the tutorials work on the newly released crates.io sources by running `cargo make test-tutorials-cratesio`

Reviewer: ```suggestion
* Merge the tutorials PR. `cargo make test-tuturials-cratesio` should now pass
```
